packages/@expo/cli/src/utils/codesigning.ts

```diff
@@ -0,0 +1,350 @@
+    if (typeof expectedAlgTemp !== 'string') {
+      throw new Error('Invalid value for alg in expo-expect-signature header');
+    }
+    expectedAlg = expectedAlgTemp;
+  }
+
+  if (expectedKeyId === 'expo-root') {
+    return await getExpoRootDevelopmentCodeSigningInfoAsync(exp);
+  } else if (expectedKeyId === 'expo-go') {
+    throw new Error('Invalid certificate requested: cannot sign with embedded keyid=expo-go key');
+  } else {
+    return await getProjectCodeSigningCertificateAsync(
+      exp,
+      privateKeyPath,
+      expectedKeyId,
+      expectedAlg
+    );
+  }
+}
+
+/**
+ * Get a development code signing certificate for the expo-root -> expo-go -> (development certificate) certificate chain.
+ * This requires the user be logged in and online, otherwise try to use the cached development certificate.
+ */
+async function getExpoRootDevelopmentCodeSigningInfoAsync(
+  exp: ExpoConfig
+): Promise<CodeSigningInfo | null> {
+  const easProjectId = exp.extra?.eas?.projectId;
+  // can't check for scope key validity since scope key is derived on the server from projectId and we may be offline.
+  // we rely upon the client certificate check to validate the scope key
+  if (!easProjectId) {
+    Log.warn('No project ID specified in app.json, unable to sign manifest');
+    return null;
+  }
+
+  // 1. If online, ensure logged in, generate key pair and CSR, fetch and cache certificate chain for projectId
+  //    (overwriting existing dev cert in case projectId changed or it has expired)
+  if (!APISettings.isOffline) {
+    return await fetchAndCacheNewDevelopmentCodeSigningInfoAsync(easProjectId);
+  }
+
+  // 2. check for cached cert/private key matching projectId and scopeKey of project, if found and valid return private key and cert chain including expo-go cert
+  const developmentCodeSigningInfoFromFile = await DevelopmentCodeSigningInfoFile.readAsync(
+    easProjectId
+  );
+  const validatedCodeSigningInfo = validateStoredDevelopmentExpoRootCertificateCodeSigningInfo(
+    developmentCodeSigningInfoFromFile,
+    easProjectId
+  );
+  if (validatedCodeSigningInfo) {
+    return validatedCodeSigningInfo;
+  }
+
+  // 3. if offline, return null
+  Log.warn('Offline and no cached development certificate found, unable to sign manifest');
+  return null;
+}
+
+/**
+ * Get the certificate configured for expo-updates for this project.
+ */
+async function getProjectCodeSigningCertificateAsync(
+  exp: ExpoConfig,
+  privateKeyPath: string | undefined,
+  expectedKeyId: string,
+  expectedAlg: string | null
+): Promise<CodeSigningInfo | null> {
+  const codeSigningCertificatePath = exp.updates?.codeSigningCertificate;
+  if (!codeSigningCertificatePath) {
+    return null;
+  }
+
+  if (!privateKeyPath) {
+    privateKeyPath = path.join(path.dirname(codeSigningCertificatePath), 'private-key.pem');
+  }
+
+  const codeSigningMetadata = exp.updates?.codeSigningMetadata;
+  if (!codeSigningMetadata) {
+    throw new Error(
```

Comment:
```suggestion
    throw new CommandError(
```
Try to utilize the `CommandError` for known and expected errors.